### PR TITLE
rarbg: implement retry strategy with 429 response

### PR DESCRIPTION
**Pull request by mynameisbogdan**
- Fixes (hopefully) #13998

### CODE REVIEW

**Comment by garfield69** [comment]
If memory serves, I believe that we do not throw an exception on a retry failure for rarbg because *arr apps will flag the indexer as not operational and stop using the indexer until the user resets manually. I may be wrong, I'm not a regular *arr apps user.

**Review by ilike2burnthing — APPROVED** [review]
Recreating the issue isn't the easiest, so I _think_ this is working fine.

I did get one error, but I'm not sure it's related:

```text
Error on 20230211025831 for RARBG.txtSystem.NullReferenceException: Object reference not set to an instance of an object.
   at Jackett.Common.Indexers.RarBG.PerformQueryWithRetry(TorznabQuery query, Boolean retry) in D:\a\1\s\src\Jackett.Common\Indexers\RarBG.cs:line 204

{"error":"No results found","error_code":20,"rate_limit":1}
```

**Comment by bakerboy448** [comment]
Why not just return a 429 back to the starrs with the correct retry after in the header?

https://github.com/Sonarr/Sonarr/blob/0a2b109a3fe101e260b623d0768240ef8b7a47ae/src/NzbDrone.Common/Http/TooManyRequestsException.cs#L20

https://github.com/Sonarr/Sonarr/blob/6660e0c3f37fade91d39141b2830dba4e20a46e2/src/NzbDrone.Core.Test/IndexerTests/NewznabTests/NewznabFixture.cs#L106

File: src/Jackett.Common/Indexers/RarBG.cs [hunk 1]
```diff
@@ -189,23 +181,13 @@ private async Task<IEnumerable<ReleaseInfo>> PerformQueryWithRetry(TorznabQuery
                 case 14: // tmdb not found (see Radarr #7625), thetvdb not found
                 case 20: // no results found
                     if (jsonContent.ContainsKey("rate_limit"))
```
Comment (bakerboy448) [inline src/Jackett.Common/Indexers/RarBG.cs:183]: only throw ratelimit if `rate_limit` is populated; HTTP is 200; and there are no results found

File: src/Jackett.Common/Indexers/RarBG.cs [hunk 2]
```diff
@@ -189,23 +181,13 @@ private async Task<IEnumerable<ReleaseInfo>> PerformQueryWithRetry(TorznabQuery
...
-                        {
-                            logger.Warn("torrentapi.org returned code 20 with Rate Limit exceeded.");
-                            return releases;
-                        }
-                    }
+                        throw new TooManyRequestsException("Rate limited, retry in 5 minutes", TimeSpan.FromMinutes(5));
+
                     // the api returns "no results" in some valid queries. we do one retry on this case but we can't do more
                     // because we can't distinguish between search without results and api malfunction
                     return retry ? await PerformQueryWithRetry(query, false) : releases;
```
Comment (bakerboy448) [inline src/Jackett.Common/Indexers/RarBG.cs:188]: this can likely be removed; no?

File: src/Jackett.Server/Controllers/ResultsController.cs [hunk 3]
```diff
@@ -440,7 +454,18 @@ public async Task<IActionResult> Torznab([FromQuery] TorznabRequest request)
         }
 
         [Route("[action]/{ignored?}")]
-        public IActionResult GetErrorXML(int code, string description) => Content(CreateErrorXML(code, description), "application/xml", Encoding.UTF8);
+        public IActionResult GetErrorXML(int code, string description, int statusCode = StatusCodes.Status400BadRequest)
```
Comment (bakerboy448) [inline src/Jackett.Server/Controllers/ResultsController.cs:457]: should the default be moved to be within the code; looks like that would possibly be more consistent?

File: src/Jackett.Common/Indexers/RarBG.cs (hunk 1 above)
Comment (bakerboy448) [reply src/Jackett.Common/Indexers/RarBG.cs:183]: the quoted block is what rarbg indicated for their rate limit handling;  but it seemed when that was implemented in prowlarr they would return a 200 with a rate_limit of 1 and yet not actually be rate limited or something?

**Comment by ilike2burnthing** [comment]
Will have to test later, currently getting rate limited errors on 90% of requests whether I'm using this build or the current release, using public IP or proxy.

**Comment by bakerboy448** [comment]
@rarbg any insight on those?

**Comment by ilike2burnthing** [comment]
Still having the same issue unfortunately. Anyone else?

**Comment by ilike2burnthing** [comment]
> 1req/5 minutes

Ouch! I got one error in ~15-20mins of testing multiple times a minute. It would take me hours to repeat that now.

Still at a loss as to why 2 days ago I was fine, only getting rate limit warnings when I forced it to happen, but since yesterday I now can't get 2 consecutive successful searches.

**Comment by ilike2burnthing** [comment]
After a bit more trial and error, I can make a request every 30s, otherwise I get 520 status errors. Not quite as as bad as 1req/5mins but still annoying. @rarbg any information I can provide to try to get this resolved, or would you prefer I email support?

As for this PR, I wasn't able to do as much testing as previously, but I'm not getting any errors (other than 520), and '0 results' from an error isn't being cached, whereas '0 results' from `abcdef123456` is. Looks good. If you're happy with it, merge.

**Comment by rarbg** [comment]
@ilike2burnthing please email support

**Comment by ilike2burnthing** [comment]
Ok, after a fair bit of troubleshooting we've narrowed down the issue to `mode=list` (keywordless search). The required delay was able to be reduced to 20s with CF cache disabled, but that's not really worth it.

@mynameisbogdan are you able to reproduce this?
- disable cache in Jackett
- perform a keywordless search with RARBG (do not use `Test`, it does not respect requestDelay), get results
- wait for ~20s, perform another keywordless search, get 520 error
- wait for another ~10s (so total >30s), perform another keywordless search, get results

To make things a little odder, while getting 520 errors for `mode=list`, you can still use `mode=search`.

If you're able to reproduce, is it possible to set requestDelay to 30s for `mode=list` but keep it at 5s for `mode=search`?

**Pull request merged**